blas/testblas/level1double.go

```diff
@@ -1772,7 +1772,7 @@ var DrotmgTests = []DrotmgTestStruct{
 		Name: "RD1_Big_RD2_Big_Flag_0",
 		P: &blas.DrotmParams{
 			Flag: blas.Rescaling,
-			H:    [4]float64{4096, -3584, 1792, 4096},
+			H:    [4]float64{4096, -4096, 1, 4096},
```

Comment:
I'm getting failures here:

```
level1double.go:2109: drotm y_0 mismatch RD1_Big_RD2_Big_Flag_0: expected 0, found -4096
```

Reply:
And I see that this is https://github.com/xianyi/OpenBLAS/issues/1452#issuecomment-366565894

Reply:
Is it expected that our tests also fail?

Reply:
The OpenBLAS implementation fails in exactly the same way that this code does. So for some value of expectation, yes, it's expected. I'd like this not to fail and for us to be able to provide a solution to the OpenBLAS failure.

Reply:
This is the only part that we need to resolve. It is related though to the new test that I have added in below (at L2104) and am not sure about; it does not look correct wrt the documentation, but is the only way I can get Drotm to zero out the y1 component. If you can help resolve that, we can probably finish this up.

Reply:
By way of notes, prior to the rescaling section of `Drotmg` the values of H would work correctly as a `blas.Rescaling` H for this input, so the issue is again in the rescaling part.

Reply:
This is the penalty we reap for making independent tests

Reply:
Can you confirm that I have interpreted the documentation of `DROTMG` correctly in the additional test stanza at [L2104](https://github.com/gonum/gonum/blob/783e09b9285a8f952d51b722de5c168a0a0c5529/blas/testblas/level1double.go#L2104-L2111).

If that's not correct, we can just remove it or fix it and maybe this failure goes away. If it is correct, it should not be too difficult to figure out what in the rescaling code is being broken.

Reply:
Thanks for the more direct pointer. I was looking this morning trying to figure it out, but did not succeed in the time I had. I'll take another stab.

Reply:
I must be wrong, because my interpretation fails all of the tests. But, if we look at this documentation https://software.intel.com/en-us/mkl-developer-reference-fortran-rotmg , it seems like the vector of interest is `x1*sqrt(d1)` and `y1*sqrt(d2)` which is loosely confirmed by the fortran documentation. Because of that, I don't see why the vector you're supplying is `x1` and `y1`

Reply:
Indeed. Mind you the Fortran documentation is pretty terrible. I also looked at the [cublas documentation](https://pyculib.readthedocs.io/en/latest/cublas.html#pyculib.blas.Blas.rotmg) which is what I was working from - this still doesn't agree with what I have here, but what I have here is the only thing that zeros out the second entry (I tried what is actually described - unless I'm wildly off the track).

Reply:
The code is certainly doing something with the terms `x1 * sqrt(d1)` and such, but it's convoluted.

Rather than read what was going on, I tried to just solve the problem myself, but I really don't get it. `H` has four degrees of freedom. If the requirement is just to zero out the second component, then there is one constraint. If there is also a requirement on the first component (say, equal to x1), then there are two constraints. Either way, this seems like an underconstrained problem . A normal givens rotation only has two elements, `c` and `s`, but that doesn't seem like what's going on here.

Reply:
I've pushed a change that reflects the documentation in the use of`Drotm`, which fails more, and also causes NaNs to be introduced when `d1` or `d2` are negative. It would be great if the original documentation were more informative. Is there a paper for this?

Reply:
I don't have a paper but there is a paragraph in Björck (2015): Numerical methods in matrix computations:
```
It is possible to rearrange the plane rotations so that only two instead of four multi-
plications per element are used and no square roots are required. These modified trans-
formations, called fast Givens transformations, are due to Gentleman [115, 1973]
and Hammarling [146, 1974]. The basic idea is to take out a scaling factor and write
G = cQ = c
1 s/c
−s/c 1


or G = s Q = s
c/s 1
−1 c/s,


(2.3.18)
depending on whether |c| > |s| or |c|≤|s|. In a product of rotations G1 ··· Gk the
scaling factors ci and si are accumulated separately. A dynamic scaling has been
suggested by Anda and Park [1, 1994]. On modern processors the gain in speed is
modest. Because of this and the nontrivial amount of monitoring needed to avoid
overflow and underflow, the usefulness of fast Givens transformations appears to be
limited and LAPACK does not make use of them.
```
The Anda and Park reference is:
Fast plane rotations with dynamic scaling. SIAM J. Matrix Anal Appl 15, 162-714
But I don't have access to it so not sure if relevant or helpful. Perhaps the older papers.

Reply:
> H has four degrees of freedom.

I don't think that's the case. Drotmg still computes a Givens rotation, just in a way that reduces the number of arithmetical operations. It still has only only degree of freedom, the angle of rotation, and one constraint, to zero out the second component.

Reply:
Regarding the failure, I think this comment https://github.com/xianyi/OpenBLAS/issues/1452#issuecomment-368057808 is right and the logic is most likely wrong.

It's easy to see it when you comment out all other tests and only log the error so that the test passes. First, H is set to zero out y1. Then rescaling in the first component is done, so flag is Rescaling. Then rescaling for the second component is done, but since flag is Rescaling, h12 is set to 1 and h21 is set to -1. This destroys the scaling in x1 and zeroing out of y1. It makes no sense.

If I had time, I would consider rewriting this based on the Appendix A in the Hopkins paper.